File: Data Structures & Algorithms/validate-parentheses/submission-7.py
class Solution:
    def isValid(self, s: str) -> bool:
        par = {
            ')': '(',
            ']': '[',
            '}': '{'
        }

        stack = []

        for p in s:
            if p in par:
                if not stack or stack[-1] != par[p]:
                    return False
                stack.pop()
            else:
                stack.append(p)
        
        return len(stack) == 0

File: Data Structures & Algorithms/baseball-game/submission-7.ts
class Solution {
    /**
     * @param {string[]} operations
     * @return {number}
     */
    calPoints(operations: string[]): number {
        const scores = []
        let total = 0

        for (const op of operations) {
            let newScore = 0
            if (op === '+') {
                newScore = scores[scores.length - 1] + scores[scores.length - 2]
                scores.push(newScore)
            } else if (op === 'D') {
                newScore = scores[scores.length - 1] * 2
                scores.push(newScore)
            } else if (op === 'C') {
                newScore = scores.pop() * -1
            } else {
                newScore = parseInt(op)
                scores.push(newScore)
            }

            total += newScore
        }

        return total
    }
}
